File: intermine/ruby/intermine/bin/benchmark_references.rb
$LOAD_PATH << File.expand_path( File.dirname(__FILE__) + '/../lib' )

require 'benchmark'
require 'intermine/service'

include Benchmark

token = "Z1a3D3U16cicCdS0T6y4bdN1SQh"
service = Service.new("localhost/intermine-test", token )
list = service.list("My-Favourite-Employees")

def do_work(managers)
    sum = managers.reduce(0) do |m, manager| 
        m + manager.department.employees.reduce(0) do |n, emp| 
            n + emp.age
        end
    end
    avg = sum / managers.reduce(0) do |m, i| 
        m + i.department.employees.size
    end
end

Benchmark.benchmark(" "*7 + CAPTION, 7, FMTSTR, ">total:", ">avg:") do |x|
    lazy_times = x.report("lazy") do
        10.times do
            managers = list.entries
            do_work(managers)
        end
    end
    some_prefetch_times = x.report("half") do
        10.times do 
            managers = list.query.select("*", "department.*").results.entries
            do_work(managers)
        end
    end

    prefetch_times = x.report("fetched") do
        10.times do 
            managers = list.query.select("*", "department.employees.*").results.entries
            do_work(managers)
        end
    end

    [lazy_times + prefetch_times + some_prefetch_times, 
        (lazy_times + prefetch_times + some_prefetch_times)/3]
end

File: intermine/python/intermine-core/bin/benchmark_references.py
import sys
import os
sys.path.insert(0, os.getcwd())

from intermine.webservice import Service
import time

s = Service("localhost/intermine-test")
lazy = s.select("Department.*")
eager = s.select("Department.*", "Department.employees.*")

def do_work(q):
    res = q.results()
    age_sum = reduce(lambda x, y: x + reduce(lambda a, b: a + b.age, y.employees, 0), res, 0)
    assert(age_sum == 5798)

if __name__ == '__main__':
    tests = {"Lazy": lazy, "Eager": eager}
    n = 10
    print "Benchmarking %d iterations of " % n + ", ".join(tests.keys())
    for label, q in tests.iteritems():
        total = 0
        for t in range(n):
            a = time.time()
            do_work(q)
            b = time.time()
            total += b - a
        print "%s: total %s, avg: %s" % (label, total, total/n)

